clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -4486,11 +4489,66 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
   if (IsIAMCU && types::isCXX(Input.getType()))
     D.Diag(diag::err_drv_clang_unsupported) << "C++ for IAMCU";
 
+  // Handle depscan.
+  if (Job.getKind() == Action::DepscanJobClass) {
+    CmdArgs.push_back("-cc1depscan");
+
+    // Pass depscan related options to cc1depscan.
+    const OptSpecifier DepScanOpts[] = {
+        options::OPT_fdepscan_EQ,
+        options::OPT_fdepscan_share_EQ,
+        options::OPT_fdepscan_share_parent,
+        options::OPT_fdepscan_share_parent_EQ,
+        options::OPT_fno_depscan_share,
+        options::OPT_fdepscan_share_stop_EQ,
+        options::OPT_fdepscan_daemon_EQ,
+        options::OPT_fdepscan_prefix_map_EQ,
+        options::OPT_fdepscan_prefix_map_sdk_EQ,
+        options::OPT_fdepscan_prefix_map_toolchain_EQ};
+
+    Args.AddAllArgs(CmdArgs, DepScanOpts);
+
+    assert(Output.isFilename() && "Depscan needs to have file output");
+    CmdArgs.push_back("-o");
+    CmdArgs.push_back(Output.getFilename());
+    CmdArgs.push_back("-cc1-args");
+  }
+
   // Invoke ourselves in -cc1 mode.
   //
   // FIXME: Implement custom jobs for internal actions.
   CmdArgs.push_back("-cc1");
 
+  // Handle full response file input.
+  if (Inputs.front().getType() == types::TY_ResponseFile) {
+    // Render response file input first.
+    assert(Inputs.size() == 1 && "Only one response file input");
+    auto &II = Inputs.front();
+    assert(II.isFilename() && "Should be response file");
+    SmallString<128> InputName("@");
+    InputName += II.getFilename();
+    CmdArgs.push_back(C.getArgs().MakeArgString(InputName));
+
+    // FIXME: Overwrite the output filename. To fix this, we need to somehow
+    // route the final output type into Depscan ConstructJob.
+    if (Output.isFilename()) {
+      CmdArgs.push_back("-o");
+      CmdArgs.push_back(Output.getFilename());
+    }
+    const char *Exec = D.getClangProgramPath();
+    if (D.CC1Main && !D.CCGenDiagnostics) {
```

Comment:
I'm fine with the code duplication since it's pretty short, but please add a FIXME for outlining into a function somehow when upstreaming.

Reply:
This is not code duplication. This is a new early return.

Reply:
I know the early return is new. The logic that takes place starting here, before the return, is non-trivial and essentially identical to the code later in the function where it was copied from. Properly factored, there'd be a lambda function for this `if` block and the call to `getClangProgramPath()`.

Reply:
Sure. I will add a note.